Profiler only logs collector connectivity issues as DEBUG

This came up in a customer troubleshooting session. No data was coming to the collector, but no errors were displayed in the agent logs. After debug logs were turned on via `-Dotel.javaagent.debug=true`, the problem became clear. As it turned out, the endpoint was specified as `https://localhost:4317` (note the protocol)

```
Caused by: io.grpc.netty.shaded.io.netty.handler.ssl.NotSslRecordException: not an SSL/TLS record: 000006040000000000000500004000
```

To make future troubleshooting faster, this sort of errors should be logged at the WARN level. To avoid polluting the logs too much, consider logging it just once per run or per hour. We should also very non-ambiguously mark the log messages as coming from the Splunk Java Agent, not from the instrumented application itself.

--- a/profiler/src/main/java/com/splunk/opentelemetry/logs/OtlpLogsExporter.java
+++ b/profiler/src/main/java/com/splunk/opentelemetry/logs/OtlpLogsExporter.java
@@ -26,11 +26,11 @@ import io.opentelemetry.proto.collector.logs.v1.ExportLogsServiceResponse;
 import io.opentelemetry.proto.collector.logs.v1.LogsServiceGrpc;
 import io.opentelemetry.proto.logs.v1.ResourceLogs;
 import io.opentelemetry.sdk.logs.data.LogData;
+import java.time.Clock;
 import java.time.Duration;
 import java.time.Instant;
 import java.util.List;
 import java.util.concurrent.atomic.AtomicReference;
-import java.util.function.Supplier;
 import javax.annotation.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -94,18 +94,18 @@ public class OtlpLogsExporter implements LogsExporter {
   @VisibleForTesting
   static class OncePerHourLogger {
     private final AtomicReference<Instant> lastErrorTime = new AtomicReference<>(Instant.EPOCH);
-    private final Supplier<Instant> clock;
+    private final Clock clock;
 
     OncePerHourLogger() {
-      this(Instant::now);
+      this(Clock.systemUTC());
     }
 
-    OncePerHourLogger(Supplier<Instant> clock) {
+    OncePerHourLogger(Clock clock) {
       this.clock = clock;
     }
 
     public void log(String message, Throwable th) {
-      Instant now = clock.get();
+      Instant now = clock.instant();
       if (Duration.between(lastErrorTime.get(), now).toHours() > 0) {
         lastErrorTime.set(now);
         logger.error(message, th);

--- a/profiler/src/main/java/com/splunk/opentelemetry/logs/OtlpLogsExporter.java
+++ b/profiler/src/main/java/com/splunk/opentelemetry/logs/OtlpLogsExporter.java
@@ -18,6 +18,7 @@ package com.splunk.opentelemetry.logs;
 
 import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
 
+import com.google.common.annotations.VisibleForTesting;
 import com.google.common.util.concurrent.FutureCallback;
 import com.google.common.util.concurrent.Futures;
 import io.opentelemetry.proto.collector.logs.v1.ExportLogsServiceRequest;
@@ -25,7 +26,11 @@ import io.opentelemetry.proto.collector.logs.v1.ExportLogsServiceResponse;
 import io.opentelemetry.proto.collector.logs.v1.LogsServiceGrpc;
 import io.opentelemetry.proto.logs.v1.ResourceLogs;
 import io.opentelemetry.sdk.logs.data.LogData;
+import java.time.Duration;
+import java.time.Instant;
 import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
 import javax.annotation.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -35,6 +40,7 @@ public class OtlpLogsExporter implements LogsExporter {
 
   private final ResourceLogsAdapter adapter;
   private final LogsServiceGrpc.LogsServiceFutureStub client;
+  private final OncePerHourLogger errorLogger = new OncePerHourLogger();
 
   public OtlpLogsExporter(
       LogsServiceGrpc.LogsServiceFutureStub client, ResourceLogsAdapter adapter) {
@@ -48,7 +54,7 @@ public class OtlpLogsExporter implements LogsExporter {
     ExportLogsServiceRequest request =
         ExportLogsServiceRequest.newBuilder().addResourceLogs(resourceLogs).build();
 
-    ResponseHandler responseHandler = new ResponseHandler(logs);
+    ResponseHandler responseHandler = new ResponseHandler(logs, errorLogger);
     Futures.addCallback(client.export(request), responseHandler, directExecutor());
   }
 
@@ -67,9 +73,11 @@ public class OtlpLogsExporter implements LogsExporter {
   private static class ResponseHandler implements FutureCallback<ExportLogsServiceResponse> {
 
     private final List<LogData> logs;
+    private final OncePerHourLogger errorLogger;
 
-    public ResponseHandler(List<LogData> logs) {
+    public ResponseHandler(List<LogData> logs, OncePerHourLogger errorLogger) {
       this.logs = logs;
+      this.errorLogger = errorLogger;
     }
 
     @Override
@@ -79,7 +87,31 @@ public class OtlpLogsExporter implements LogsExporter {
 
     @Override
     public void onFailure(Throwable th) {
-      logger.debug("Failed to export logs", th);
+      errorLogger.log("Splunk profiling agent failed to export logs", th);
+    }
+  }
+
+  @VisibleForTesting
+  static class OncePerHourLogger {
+    private final AtomicReference<Instant> lastErrorTime = new AtomicReference<>(Instant.EPOCH);
+    private final Supplier<Instant> clock;
+
+    OncePerHourLogger() {
+      this(Instant::now);
+    }
+
+    OncePerHourLogger(Supplier<Instant> clock) {
+      this.clock = clock;
+    }
+
+    public void log(String message, Throwable th) {
+      Instant now = clock.get();
+      if (Duration.between(lastErrorTime.get(), now).toHours() > 0) {
+        lastErrorTime.set(now);
+        logger.error(message, th);
+      } else {
+        logger.debug(message, th);
+      }
     }
   }
 }

--- a/profiler/src/main/java/com/splunk/opentelemetry/logs/OtlpLogsExporter.java
+++ b/profiler/src/main/java/com/splunk/opentelemetry/logs/OtlpLogsExporter.java
@@ -26,11 +26,7 @@ import io.opentelemetry.proto.collector.logs.v1.ExportLogsServiceResponse;
 import io.opentelemetry.proto.collector.logs.v1.LogsServiceGrpc;
 import io.opentelemetry.proto.logs.v1.ResourceLogs;
 import io.opentelemetry.sdk.logs.data.LogData;
-import java.time.Clock;
-import java.time.Duration;
-import java.time.Instant;
 import java.util.List;
-import java.util.concurrent.atomic.AtomicReference;
 import javax.annotation.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -40,7 +36,6 @@ public class OtlpLogsExporter implements LogsExporter {
 
   private final ResourceLogsAdapter adapter;
   private final LogsServiceGrpc.LogsServiceFutureStub client;
-  private final OncePerHourLogger errorLogger = new OncePerHourLogger();
 
   public OtlpLogsExporter(
       LogsServiceGrpc.LogsServiceFutureStub client, ResourceLogsAdapter adapter) {
@@ -54,7 +49,7 @@ public class OtlpLogsExporter implements LogsExporter {
     ExportLogsServiceRequest request =
         ExportLogsServiceRequest.newBuilder().addResourceLogs(resourceLogs).build();
 
-    ResponseHandler responseHandler = new ResponseHandler(logs, errorLogger);
+    ResponseHandler responseHandler = new ResponseHandler(logs);
     Futures.addCallback(client.export(request), responseHandler, directExecutor());
   }
 
@@ -70,14 +65,13 @@ public class OtlpLogsExporter implements LogsExporter {
     return new OtlpLogsExporterBuilder();
   }
 
-  private static class ResponseHandler implements FutureCallback<ExportLogsServiceResponse> {
+  @VisibleForTesting
+  static class ResponseHandler implements FutureCallback<ExportLogsServiceResponse> {
 
     private final List<LogData> logs;
-    private final OncePerHourLogger errorLogger;
 
-    public ResponseHandler(List<LogData> logs, OncePerHourLogger errorLogger) {
+    public ResponseHandler(List<LogData> logs) {
       this.logs = logs;
-      this.errorLogger = errorLogger;
     }
 
     @Override
@@ -87,31 +81,7 @@ public class OtlpLogsExporter implements LogsExporter {
 
     @Override
     public void onFailure(Throwable th) {
-      errorLogger.log("Splunk profiling agent failed to export logs", th);
-    }
-  }
-
-  @VisibleForTesting
-  static class OncePerHourLogger {
-    private final AtomicReference<Instant> lastErrorTime = new AtomicReference<>(Instant.EPOCH);
-    private final Clock clock;
-
-    OncePerHourLogger() {
-      this(Clock.systemUTC());
-    }
-
-    OncePerHourLogger(Clock clock) {
-      this.clock = clock;
-    }
-
-    public void log(String message, Throwable th) {
-      Instant now = clock.instant();
-      if (Duration.between(lastErrorTime.get(), now).toHours() > 0) {
-        lastErrorTime.set(now);
-        logger.error(message, th);
-      } else {
-        logger.debug(message, th);
-      }
+      logger.warn("Splunk profiling agent failed to export logs", th);
     }
   }
 }

--- a/profiler/src/test/java/com/splunk/opentelemetry/logs/OtlpLogsExporterTest.java
+++ b/profiler/src/test/java/com/splunk/opentelemetry/logs/OtlpLogsExporterTest.java
@@ -1,0 +1,51 @@
+/*
+ * Copyright Splunk Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.splunk.opentelemetry.logs;
+
+import static java.time.temporal.ChronoUnit.HOURS;
+import static java.time.temporal.ChronoUnit.MINUTES;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import io.github.netmikey.logunit.api.LogCapturer;
+import java.time.Instant;
+import java.util.function.Supplier;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+class OtlpLogsExporterTest {
+
+  @RegisterExtension LogCapturer log = LogCapturer.create().captureForType(OtlpLogsExporter.class);
+
+  @Test
+  void oncePerHourLogging_debugDisabled() {
+    Supplier<Instant> clock = mock(Supplier.class);
+    Instant firstTime = Instant.now();
+    Instant secondTime = firstTime.plus(45, MINUTES);
+    Instant thirdTime = firstTime.plus(1, HOURS);
+    when(clock.get()).thenReturn(firstTime, secondTime, thirdTime);
+
+    OtlpLogsExporter.OncePerHourLogger perHour = new OtlpLogsExporter.OncePerHourLogger(clock);
+    perHour.log("test1", null);
+    perHour.log("test2", null); // should skip
+    perHour.log("test3", null); // should register
+
+    log.assertContains("test1");
+    log.assertDoesNotContain("test2");
+    log.assertContains("test3");
+  }
+}

--- a/profiler/src/test/java/com/splunk/opentelemetry/logs/OtlpLogsExporterTest.java
+++ b/profiler/src/test/java/com/splunk/opentelemetry/logs/OtlpLogsExporterTest.java
@@ -22,8 +22,8 @@ import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
 
 import io.github.netmikey.logunit.api.LogCapturer;
+import java.time.Clock;
 import java.time.Instant;
-import java.util.function.Supplier;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
@@ -33,11 +33,11 @@ class OtlpLogsExporterTest {
 
   @Test
   void oncePerHourLogging_debugDisabled() {
-    Supplier<Instant> clock = mock(Supplier.class);
+    Clock clock = mock(Clock.class);
     Instant firstTime = Instant.now();
     Instant secondTime = firstTime.plus(45, MINUTES);
     Instant thirdTime = firstTime.plus(1, HOURS);
-    when(clock.get()).thenReturn(firstTime, secondTime, thirdTime);
+    when(clock.instant()).thenReturn(firstTime, secondTime, thirdTime);
 
     OtlpLogsExporter.OncePerHourLogger perHour = new OtlpLogsExporter.OncePerHourLogger(clock);
     perHour.log("test1", null);

--- a/profiler/src/test/java/com/splunk/opentelemetry/logs/OtlpLogsExporterTest.java
+++ b/profiler/src/test/java/com/splunk/opentelemetry/logs/OtlpLogsExporterTest.java
@@ -16,14 +16,9 @@
 
 package com.splunk.opentelemetry.logs;
 
-import static java.time.temporal.ChronoUnit.HOURS;
-import static java.time.temporal.ChronoUnit.MINUTES;
-import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.when;
+import static java.util.Collections.emptyList;
 
 import io.github.netmikey.logunit.api.LogCapturer;
-import java.time.Clock;
-import java.time.Instant;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
@@ -32,20 +27,9 @@ class OtlpLogsExporterTest {
   @RegisterExtension LogCapturer log = LogCapturer.create().captureForType(OtlpLogsExporter.class);
 
   @Test
-  void oncePerHourLogging_debugDisabled() {
-    Clock clock = mock(Clock.class);
-    Instant firstTime = Instant.now();
-    Instant secondTime = firstTime.plus(45, MINUTES);
-    Instant thirdTime = firstTime.plus(1, HOURS);
-    when(clock.instant()).thenReturn(firstTime, secondTime, thirdTime);
-
-    OtlpLogsExporter.OncePerHourLogger perHour = new OtlpLogsExporter.OncePerHourLogger(clock);
-    perHour.log("test1", null);
-    perHour.log("test2", null); // should skip
-    perHour.log("test3", null); // should register
-
-    log.assertContains("test1");
-    log.assertDoesNotContain("test2");
-    log.assertContains("test3");
+  void exportLogs() {
+    OtlpLogsExporter.ResponseHandler handler = new OtlpLogsExporter.ResponseHandler(emptyList());
+    handler.onFailure(new RuntimeException("kaboom"));
+    log.assertContains("failed to export logs");
   }
 }

